ADR/00xx-component-dependencies.md

```diff
@@ -0,0 +1,171 @@
+
+The [build-service] will also update the PR it filed when the PRs that triggered it are merged or updated. It may update the description and/or it may rebase on the main branch and/or it may issue /retest. Need to figure out what we want here.
+
+### Integration-service and PR Groups
+
+* When [integration-service] notices a build from a PR (PR#2) that it detects is in the same group as a build from another PR (PR#1), it will perform testing but it will perform that process with some modifications. It will:
+  * Construct the Snapshot for the test using the image from the triggering build pipeline of PR #2 as well as image found on the latest build pipeline associated with PR#1. This lets a user test PR #2 including unmerged content from PR #1. It will construct the Snapshot from the latest builds of all PRs in the PR Group.
+  * Post the test results back on PR#1 in addition to the results it would normally post to PR#2. It will post the test results back to all PRs in the PR Group.
+
+### Build-service and PR Groups
+
+* When [build-service] responds to the build of a PR (PR #1) and propagates the digest from one Component to another as a PR (PR#2). It follows the declared `depends on` references to know which other repos should receive an update PR. It marks the PR that is submits (PR #2) as being in the same PR Group as the triggering PR (PR #1). This enables pre-merge testing of both changes.
+* When [build-service] submits PR #2, it marks it as "Draft" and it includes a reference to the triggering PR (PR #1) in the description of the automatically submitted PR (PR #2), giving the user some indication that it should not be merged out of order.
+
+## Applied to Use Cases
+
+Let's apply the architecture to some use cases, and see how it plays out:
+
+### Common Parent Image
+
+Scenario: an application image depends on a common parent image. The user has 1 Application, and 2 Components. One of them is the parent image. The other image is built FROM that image.
+
+* The child image Component declares that it `depends on` the parent image Component, by way of a the new field on the Component CR.
+* [integration-service] will always skip testing for parent image update builds, will never promote them, or use them to initiate Releases, but it will promote them to the global candidate list.
+* [build-service] will propagate digest references as a PR to the child image Component repo, by analyzing the `depends on` fields of all other Components in the Application. The PR that it files must be submitted in such a way that it appears in the same PR Group as the triggering PR submitted by a user.
+* When the parent image PR is merged, [build-service] will update the PRs it originally filed to take them out of "Draft" to indicate that they are safe to merge now as long as there are no other unmerged triggering PRs in the same group. It may potentially rebase the PRs to trigger a new build or use `/retest`.
```

Comment:
> So when the parent image change is merged, no release is triggered, but it is released when the child image merges and is released (assuming automatic configuration)?

Yes - because earlier when the parent image change was merged - it got promoted to the global candidate list. This means that later when the child image merges, it prompts the creation of a new Snapshot using normal rules: the Snapshot is constructed from the new image (child image) plus the current state of the global candidate list (which happens to include the previously merged parent).

> If a team manually triggers a release before all child image changes are also merged, is that a problem?

It depends. It could be, if the team decides it is. It could not be, if that kind of consistency isn't important to them.
If it is important to them, we can recommend they add an integration test scenario which checks for "snapshot consistency" (an idea that @arewm has been advocating), to prevent releases until we have a Snapshot which is internally consistent.

> Is there any desire in the future to replicate intermediate images which are leveraged, but never publicly shipped?

How do you mean?

I think that teams can do this by declaring a Component for their intermediate image and declare component references so that it is used as part of a build chain, but omit it explicitly from the ReleasePlan/Admission so that it doesn't get released.

Reply:
> I think that teams can do this by declaring a Component for their intermediate image and declare component references so that it is used as part of a build chain, but omit it explicitly from the ReleasePlan/Admission so that it doesn't get released.
Fair point, seems like they could be replicated if necessary. I was wondering if that would need to be a special dependency use case if we pulled that concept forward in the future.

> It depends. It could be, if the team decides it is. It could not be, if that kind of consistency isn't important to them.
If it is important to them, we can recommend they add an integration test scenario which checks for "snapshot consistency" (an idea that @arewm has been advocating), to prevent releases until we have a Snapshot which is internally consistent.
This is a tangent now, and obviously we hope teams utilize integration tests, but without any, is there value in trying to deploy their app to an ephemeral environment before deploying to their first static environment? Running the default contract check as we have it today wouldn't help us understand if the snapshot couldn't be deployed, right?

Reply:
What would the criteria be for the deployment to ephemeral environments? If there are no remaining related components that need to be resolved, then the snapshot _should_ be consistent internally (in the related PRs). The eventual consistency of the snapshots is for release-time verification.

If we want to take into account eventual consistency of the snapshots for push builds in integration tests, then an EC check can be done (and potentially fail) before an integration test is completed.